datalad/support/gitrepo.py

```diff
@@ -2300,211 +2392,231 @@ def fetch(self, remote=None, refspec=None, all_=False, **kwargs):
-
-        Returns
-        -------
-        list
-            FetchInfo objects of the items fetched from remote
+        git_options : list, optional
+          Additional command line options for git-push.
         """
-        # TODO: options=> **kwargs):
-        # Note: Apparently there is no explicit (fetch --all) in gitpython,
-        #       but fetch is always bound to a certain remote instead.
-        #       Therefore implement it on our own:
-        if remote is None:
-            if refspec is not None:
-                # conflicts with using tracking branch or fetch all remotes
-                # For now: Just fail.
-                # TODO: May be check whether it fits to tracking branch
-                raise ValueError("refspec specified without a remote. (%s)" %
-                                 refspec)
-            if all_:
-                remotes_to_fetch = [
-                    self.repo.remote(r)
-                    for r in self.get_remotes(with_urls_only=True)
-                ]
-            else:
-                # No explicit remote to fetch.
-                # => get tracking branch:
-                tb_remote, refspec = self.get_tracking_branch()
-                if tb_remote is not None:
-                    remotes_to_fetch = [self.repo.remote(tb_remote)]
-                else:
-                    # No remote, no tracking branch
-                    # => fail
-                    raise ValueError("Neither a remote is specified to fetch "
-                                     "from nor a tracking branch is set up.")
-        else:
-            remotes_to_fetch = [self.repo.remote(remote)]
-
-        fi_list = []
-        for rm in remotes_to_fetch:
-            fetch_url = \
-                self.config.get('remote.%s.fetchurl' % rm.name,
-                                self.config.get('remote.%s.url' % rm.name,
-                                                None))
-            if fetch_url is None:
-                lgr.debug("Remote %s has no URL", rm)
-                return []
-
-            fi_list += self._call_gitpy_with_progress(
-                "Fetching %s" % rm.name,
-                rm.fetch,
-                rm.repo,
-                refspec,
-                fetch_url,
-                **kwargs
-            )
-
-        # TODO: fetch returns a list of FetchInfo instances. Make use of it.
-        return fi_list
+        yield from self._fetch_push_helper(
+            base_cmd=['git', 'fetch', '--verbose', '--progress'],
+            action='fetch',
+            urlvars=('remote.{}.url', 'remote.{}.url'),
+            protocol=GitProgress,
+            info_cls=FetchInfo,
+            info_from=1,
+            add_remote=False,
+            remote=remote,
+            refspec=refspec,
+            all_=all_,
+            git_options=git_options)
+
+    # XXX Consider removing this method. It is only used in `update()`,
+    # where it could be easily replaced with fetch+merge
```

Comment:
It's no longer used in `update()` after gh-4118.

Reply:
:1st_place_medal:

Reply:
gh-4132